src/Storages/StorageAzureBlob.cpp

```diff
@@ -969,7 +969,6 @@ RelativePathWithMetadata StorageAzureBlobSource::Iterator::next()
         }
 
         size_t current_index = index.fetch_add(1, std::memory_order_relaxed);
-
```

Comment:
Let's add for the code below 

```
std::lock_guard lock(next_mutex);
return (*blobs_with_metadata)[current_index];
```

a check right after taking the lock 
```
if (current_index >= blobs_with_metadata.size())
    throw Exception(ErrorCodes::LOGICAL_ERROR, "...");
```
?
Because as I see by sanitizer report - this is the place where we got heap-buffer-overflow. So better to protect here just in case.

Reply:
I think, to robustly fix the issue, we need to initialize `current_index` when we initialize `blobs_with_metadata` under `next_mutex`. I have an impression that unless we do that, we still can get here with index bigger that `blobs_wtih_metadata.size() - 1`. wdyt?

Reply:
Maybe I don't understand smth, but I think we still just need to place all logic about getting next batch in else branch under `next_mutex`, otherwise there a lot of possible race conditions. For example, several threads can get `need_new_batch=true` and try to get new batch and update `blobs_with_metadata` simultaniously with different batches, so we can skip some data, we can mix up indexes, we can get a race condition on inserting into `outer_blobs`, etc.
Also there is a bug when we have `filter_ast`, we reset `blob_path_with_globs` instead of `blobs_with_metadata`, so when we do `blobs_with_metadata->emplace_back(...)` `blobs_with_metadata` is actually `std::nullopt`. I tested it locally, query with `where _file = ...` crashes the server. Please, add tests with filter ast

Reply:
IMHO, I would make 2 separate iterators, one for regular keys, one for key with globs. Like it's done in StorageS3/URL/HDFS

Reply:
Separated functionality of Iterator into GlobIterator and KeysIterator. In GlobIterator extended lock to the scope of next function.

> there is a bug when we have filter_ast
@Avogar Good catch. Addressed it and added a test for the same